feat(edge): core contracts (#1265)
maticnetwork/matic-docs: docs/edge/contracts/validator-set.md

- `governance`: the address of the governance contract that will own
  the `RootValidatorSet` contract after it is initialized
- `newCheckpointManager`: the address of the `CheckpointManager` for the contract
- `validatorAddresses`: an array of validator addresses to seed the contract with
- `validatorPubkeys`: an array of validator pubkeys to seed the contract with

The function has a require statement that checks that the length of
`validatorAddresses` is equal to the length of the `validatorPubkeys`.
If these are not the same length, the function reverts with an error message.

The function then sets the value of the `checkpointManager` to the
`newCheckpointManager` argument. It then iterates over `validatorAddresses`
and registers each address as a validator. It does this by creating a
new `Validator` and storing it in the `validators` mapping using the next available
validator `id` as the `key`. The `validatorIdByAddress` mapping is also updated
to store the validator `id` corresponding to the validator `address`.

After all the validators have been registered, `currentValidatorId`
is updated to the highest validator `id` that was assigned.

> Finally, contract ownership is transfered to the `governance` contract that

Reviewer: ```suggestion
Finally, contract ownership is transferred to the `governance` contract that
```

was passed in as an argument.

### addValidators

```js
function addValidators(Validator[] calldata newValidators) external {
    require(msg.sender == checkpointManager, "ONLY_CHECKPOINT_MANAGER");
    uint256 length = newValidators.length;
    uint256 currentId = currentValidatorId;

    for (uint256 i = 0; i < length; i++) {
        validators[i + currentId + 1] = newValidators[i];

        emit NewValidator(currentId + i + 1, newValidators[i]._address, newValidators[i].blsKey);
    }

    currentValidatorId += length;
}
```

This function allows the `checkpointManager` to add new validators to the contract.

> It takes in an array of `Validator`s and iterates over them, adding each one to

Reviewer: ```suggestion
It takes in an array of Validators and iterates over them, adding each one to
```

### getValidators

```js
function getValidator(uint256 id) external view returns (Validator memory) {
    return validators[id];
}
```

This function allows anyone to retrieve the `Validator` struct for a
specific validator by `id`. It takes in an `id` as an argument and returns the
corresponding `Validator` struct from the `validators` mapping.

### getValidatorBlsKey

```js
function getValidatorBlsKey(uint256 id) external view returns (uint256[4] memory) {
    return validators[id].blsKey;
}
```

This function is a convenience function that allows anyone to retrieve the
BLS public key of a specific validator by `id`. It takes in an `id` as an
argument and returns the `blsKey` stored in the `Validator` struct for that
`id`.

### getValidatorId

```js
function getValidatorId(address address) external view returns (uint256) {
    return validatorIdByAddress[address];
}
```

This function allows anyone to retrieve the `id` of a validator by their `address`.
It takes in an `address` as an argument and returns the corresponding
`id` from the `validatorIdByAddress` mapping.

### getValidatorCount

```js
function getValidatorCount() external view returns (uint256) {

>     return currentValid

Reviewer: ```suggestion
    return currentValid;
```

>     return currentValid;

Reviewer: ```suggestion
    return currentValid;
}
```
